keps/sig-cluster-lifecycle/draft-20180316-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,243 @@
+* `kubeadm phase` function will be more visible to final users; in fact phases will
+  be "first class" elements of `kubeadm init`  user interface, not anymore a feature
+  implemented somewhere else.
+* Cleanup of kubeadm codebase (most of the `k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases`
+   package could potentially be deleted).
+
+#### Phased Command Builder
+
+How to customize the  `kubeadm init` for supporting phases without collapsing all
+the complexity of the phases management into `k8s.io/kubernetes/cmd/kubeadm/app/cmd/init.go`?.
+
+The proposed approach is to implement a PhasedCommand builder, a separated and
+reusable package in charge of all the logic required for creating and managing a
+cobra.Command capable of executing phases.
+
+`kubeadm init`  will be the first command created using such builder; the resulting
+code in `kubeadm/app/cmd/init.go` is expected to require only the definition of phases
+that are part of the init workflow and their implementation (all the scaffolding
+code will be in in the Phased Command Builder)
+
+As a bonus, the PhasedCommand builder could be re-used in future for exposing with
+the same approach any other kubeadm workflows like `kubeadm join` or `kubeadm update`.
+
+#### cmdUtil factory
+
+As of today the init workflow uses a set of accessory components that are created
+once and then used by many phases, like e.g. the client for accessing the kubernetes
+cluster API.
+
+In the current implementation of `kubeadm init` and  `kubeadm phase`  the code
+required for creating such components is duplicated n-times.
+
+In order to avoid such duplication of code and the related maintenance effort, we
+are going to create a new factory object with the responsibility to create accessory
+components, and once components are created, to share components instances across
+phases.
+
+In future/by opportunity, following additional responsibilities could be assigned
+to the cmdUtil factory:
+
+* To centralize the code for other tasks which are duplicates n-times, like e.g.
+  the validation of the masterConfiguration file with all the input parameters.
+* To collect key output information of phases that should be part of the future
+  kubeadm machine readable output (e.g. the token or the hash).
+  NB. the implementation of the kubeadm machine readable interface is out of
+  scope of this proposal, mostly because currently there is not a clean separation
+  of UX and API in  kubeadm (e.g. the code that implement phases writes to stdout)
+
+## Graduation Criteria
+
+* To create a periodic E2E test that bootstraps a cluster using phases
+* To document the new user interface for phases in kubernetes.io
+
+## Implementation History
+
+* Prototype implementation (in progress)
+
+## Drawbacks
+
+By merging phases into `kubeadm init` derives a reduced capability to customize
+the user interface for each phase. More specifically:
+
+* It would not be possible to provide any kind of advice to the user about which
```

Comment:
is there no way to have `kubeadm init certs --help` spit out different help than just the same as `kubeadm init`?

Reply:
If [phases] are defined as argument of the `kubeadm init` command as discussed in this proposal, I don't think this is feasible. The only way I'm aware of for doing his is to maintain a separate three of command for phases (see alternatives)

Reply:
I'm only starting to get familiar with the codebase, but could we use [`SetHelpFunc`](https://github.com/spf13/cobra/blob/master/command.go#L223)? It looks like the second argument is the positional arguments, so we could switch based on that.

Reply:
Unfortunately `setHelpFunc` seems a dead end; [here](https://github.com/spf13/cobra/blob/a1f051bc3eba734da4772d60e2d677f47cf93ef4/command.go#L322) arguments are not effectively passed, and [here](https://github.com/spf13/cobra/blob/a1f051bc3eba734da4772d60e2d677f47cf93ef4/command.go#L851) argument are passed but only in case the CMD is executed and fails...

However, apart from the technical solution, the real question is what kind of dedicated help we think is necessary for each function, and the effort for maintaining this for about thirty phases.

If I look at current phases implementation, It seems to me that the real added value of custom help is low, so I'm basically fine in getting rid of the maintenance of all those text and of the tedious task of keeping aligned flags between phases and init.

Reply:
I'm looking over the existing phases and it seems like there's a pretty good number of phase- and subphase-specific arguments. I guess the question is whether individual users would get much help out of breaking them out. I would lean towards having _some_ machine-readable summary of which phases and subphases take which arguments. In the future we could use those to make sure our test coverage is complete over valid combinations of arguments.